[BUG] AzureSearchWriter sends GeographyPoint field as JSON string instead of GeoJSON object, causing Azure AI Search request failure.

### SynapseML version

synapseml_2.12:1.0.13

### System information

- **Language version** (e.g. python 3.8, scala 2.12): python 3.12
- **Spark Version** (e.g. 3.2.3): 3.4.4, 3.5.1
- **Spark Platform** (e.g. Synapse, Databricks): Naked, Databricks


### Describe the problem

When using AzureSearchWriter in SynapseML to update a GeographyPoint field in Azure AI Search, the writer sends the location value as a JSON string rather than a proper GeoJSON object. Azure AI Search expects a valid spatial object, so the request fails with HTTP 400.

### Code to reproduce issue

```python
from pyspark.sql import SparkSession, functions as F, types as T
import synapse.ml.services.search 

AZURE_SEARCH_ADMIN_KEY = "api_key"
AZURE_SEARCH_SERVICE   = "ai-search"
INDEX_NAME             = "demo-geoindex"

spark = (
    SparkSession.builder
    .appName("synapseml-to-azure-search")
    .config("spark.jars.packages", "com.microsoft.azure:synapseml_2.12:1.0.13")
    .getOrCreate()
)

indexJson = r'''
{
  "name": "demo-geoindex",
  "fields": [
    { "name": "id", "type": "Edm.String", "key": true, "searchable": true, "retrievable": true },
    { "name": "location", "type": "Edm.GeographyPoint", "searchable": false, "filterable": true, "retrievable": true, "sortable": true }
  ]
}
'''.strip()

# location as STRING containing GeoJSON (lon, lat)
doc_schema = T.StructType([
    T.StructField("id", T.StringType(), False),
    T.StructField("location", T.StringType(), False),
])

geojson = '{"type":"Point","coordinates":[-0.1276, 51.5072]}'

df = (spark.createDataFrame([("1", geojson)], schema=doc_schema)
          .withColumn("searchAction", F.lit("upload")))

df.show(1)

df.writeToAzureSearch(
    subscriptionKey=AZURE_SEARCH_ADMIN_KEY,
    serviceName=AZURE_SEARCH_SERVICE,
    indexJson=indexJson,
    actionCol="searchAction",
    batchSize="1"
)
```

UTILITY - REPRO without Spark:
```shell
curl -s \
 "https://<ai-search>.search.windows.net/indexes/demo-geoindex/docs/index?api-version=2023-07-01-Preview" \
 -H "api-key: <api_key>" \
    -H "Content-Type: application/json" \
    -d '{"value":[{"id":"1","location":"{\"type\":\"Point\",\"coordinates\":[-0.1276, 51.5072]}","@search.action":"upload"}]}'
```

UTILITY - EXPECTED
```shell
curl -s \
 "https://<ai-search>.search.windows.net/indexes/demo-geoindex/docs/index?api-version=2023-07-01-Preview" \
 -H "api-key: <api_key>" \
    -H "Content-Type: application/json" \
    -d '{"value":[{"id":"3","location":{"type":"Point","coordinates":[-0.1276, 51.5072]},"@search.action":"upload"}]}'
```

### Other info / logs

```log
25/09/16 19:15:04 WARN HandlingUtils: got error  400: Bad Request on https://ai-search-m-se.search.windows.net/indexes/demo-geoindex/docs/index?api-version=2023-07-01-Preview   {"value":[{"id":"1","location":"{\"type\":\"Point\",\"coordinates\":[-0.1276, 51.5072]}","@search.action":"upload"}]}
25/09/16 19:15:04 INFO SynapseMLLogging: finished sending to https://ai-search-m-se.search.windows.net/indexes/demo-geoindex/docs/index?api-version=2023-07-01-Preview took (248ms)
25/09/16 19:15:04 INFO CodeGenerator: Code generated in 13.930916 ms
25/09/16 19:15:04 INFO CodeGenerator: Code generated in 11.4245 ms
25/09/16 19:15:04 INFO CodeGenerator: Code generated in 20.387 ms
25/09/16 19:15:04 INFO CodeGenerator: Code generated in 6.13 ms
25/09/16 19:15:04 ERROR Executor: Exception in task 10.0 in stage 3.0 (TID 21)
org.apache.spark.SparkException: [FAILED_EXECUTE_UDF] Failed to execute user defined function (functions$$$Lambda$3203/0x000000700207c000: (struct<response:string,status:struct<protocolVersion:struct<protocol:string,major:int,minor:int>,statusCode:int,reasonPhrase:string>>, struct<col1:string>) => struct<response:string,status:struct<protocolVersion:struct<protocol:string,major:int,minor:int>,statusCode:int,reasonPhrase:string>>).
	at org.apache.spark.sql.errors.QueryExecutionErrors$.failedExecuteUserDefinedFunctionError(QueryExecutionErrors.scala:217)
	at org.apache.spark.sql.errors.QueryExecutionErrors.failedExecuteUserDefinedFunctionError(QueryExecutionErrors.scala)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.ScalaUDF_0$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.writeFields_0_0$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.apply(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.apply(Unknown Source)
	at scala.collection.Iterator$$anon$10.next(Iterator.scala:461)
	at scala.collection.Iterator$$anon$10.next(Iterator.scala:461)
	at scala.collection.Iterator$$anon$10.next(Iterator.scala:461)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at org.apache.spark.rdd.RDD.$anonfun$foreach$2(RDD.scala:1004)
	at org.apache.spark.rdd.RDD.$anonfun$foreach$2$adapted(RDD.scala:1004)
	at org.apache.spark.SparkContext.$anonfun$runJob$5(SparkContext.scala:2298)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:92)
	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:161)
	at org.apache.spark.scheduler.Task.run(Task.scala:139)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:554)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1529)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:557)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.RuntimeException: Service Exception:
	 [{"error":{"code":"","message":"The request is invalid. Details: The value specified for the spatial property was not valid. You must specify a valid spatial value."}},[[HTTP,1,1],400,Bad Request]]
 for input:
	 [{"value":[{"id":"1","location":"{\"type\":\"Point\",\"coordinates\":[-0.1276, 51.5072]}","@search.action":"upload"}]}]
	at com.microsoft.azure.synapse.ml.services.search.AzureSearchWriter$.$anonfun$checkForErrors$1(AzureSearch.scala:156)
	at scala.Option.map(Option.scala:230)
	at com.microsoft.azure.synapse.ml.services.search.AzureSearchWriter$.checkForErrors(AzureSearch.scala:153)
	at com.microsoft.azure.synapse.ml.services.search.AzureSearchWriter$.$anonfun$prepareDF$23(AzureSearch.scala:356)
	at org.apache.spark.injections.UDFUtils$$anon$2.call(UDFUtils.scala:29)
	at org.apache.spark.sql.functions$.$anonfun$udf$93(functions.scala:5438)
	... 22 more
```

### What component(s) does this bug affect?

- [x] `area/cognitive`: Cognitive project
- [ ] `area/core`: Core project
- [ ] `area/deep-learning`: DeepLearning project
- [ ] `area/lightgbm`: Lightgbm project
- [ ] `area/opencv`: Opencv project
- [ ] `area/vw`: VW project
- [ ] `area/website`: Website
- [ ] `area/build`: Project build system
- [ ] `area/notebooks`: Samples under notebooks folder
- [ ] `area/docker`: Docker usage
- [ ] `area/models`: models related issue

### What language(s) does this bug affect?

- [x] `language/scala`: Scala source code
- [x] `language/python`: Pyspark APIs
- [ ] `language/r`: R APIs
- [ ] `language/csharp`: .NET APIs
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [x] `integrations/synapse`: Azure Synapse integrations
- [ ] `integrations/azureml`: Azure ML integrations
- [x] `integrations/databricks`: Databricks integrations

--- a/cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/search/AzureSearch.scala
+++ b/cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/search/AzureSearch.scala
@@ -20,7 +20,8 @@ import org.apache.spark.ml.util._
 import org.apache.spark.ml.{ComplexParamsReadable, NamespaceInjections, PipelineModel}
 import org.apache.spark.ml.linalg.SQLDataTypes.VectorType
 import org.apache.spark.ml.functions.vector_to_array
-import org.apache.spark.sql.functions.{col, expr, struct, to_json, to_utc_timestamp, date_format, when}
+import org.apache.spark.sql.functions.{col, expr, from_json, struct, to_json, to_utc_timestamp,
+  date_format, when}
 import org.apache.spark.sql.streaming.DataStreamWriter
 import org.apache.spark.sql.types._
 import org.apache.spark.sql.{DataFrame, Dataset, Row}
@@ -249,6 +250,50 @@ object AzureSearchWriter extends IndexParser with IndexJsonGetter with SLogging 
     }
   }
 
+  /**
+   * Converts string columns containing GeoJSON to the proper struct shape required for
+   * Azure Search `Edm.GeographyPoint` fields.
+   *
+   * Azure AI Search expects spatial values to be sent as a GeoJSON object
+   * (e.g. `{"type":"Point","coordinates":[lon, lat]}`), not as a JSON-encoded string.
+   * Users frequently have their GeoJSON readily available as a string column, and
+   * passing it as a `StringType` previously caused a `400 Bad Request`
+   * (see issue #2420) because the writer JSON-escaped the entire string.
+   *
+   * For each field declared as `Edm.GeographyPoint` in the index, if the corresponding
+   * DataFrame column is a `StringType`, parse it into the canonical
+   * `struct<type:string, coordinates:array<double>>` so that downstream `to_json`
+   * emits a proper GeoJSON object. Columns that are already structured are left as-is.
+   *
+   * @param df DataFrame with potential GeographyPoint columns
+   * @param indexJson JSON string containing the index schema
+   * @return DataFrame with string GeographyPoint columns converted to GeoJSON structs
+   */
+  private def convertGeographyPointToStruct(df: DataFrame, indexJson: String): DataFrame = {
+    val geoStructType = StructType(Seq(
+      StructField("type", StringType),
+      StructField("coordinates", ArrayType(DoubleType))
+    ))
+    val geoFields = parseIndexJson(indexJson).fields
+      .filter(_.`type` == "Edm.GeographyPoint")
+      .map(_.name)
+    geoFields.foldLeft(df) { (currentDF, fieldName) =>
+      if (currentDF.columns.contains(fieldName)) {
+        currentDF.schema(fieldName).dataType match {
+          case StringType =>
+            currentDF.withColumn(fieldName,
+              when(col(fieldName).isNotNull, from_json(col(fieldName), geoStructType))
+            )
+          case _ =>
+            // Already a struct (or otherwise compatible) — let checkSchemaParity validate it.
+            currentDF
+        }
+      } else {
+        currentDF
+      }
+    }
+  }
+
   private def dfToIndexJson(schema: StructType,
                             indexName: String,
                             keyCol: String,
@@ -328,17 +373,18 @@ object AzureSearchWriter extends IndexParser with IndexJsonGetter with SLogging 
 
     SearchIndex.createIfNoneExists(subscriptionKey, serviceName, indexJson, apiVersion)
     val dateConvertedDF = convertDateTimeToISO8601(preppedDF, indexJson)
+    val geoConvertedDF = convertGeographyPointToStruct(dateConvertedDF, indexJson)
 
     logInfo("checking schema parity")
-    checkSchemaParity(dateConvertedDF.schema, indexJson, actionCol)
+    checkSchemaParity(geoConvertedDF.schema, indexJson, actionCol)
 
     val df1 = if (filterNulls) {
       val collectionColumns = parseIndexJson(indexJson).fields
         .filter(_.`type`.startsWith("Collection"))
         .map(_.name)
-      collectionColumns.foldLeft(dateConvertedDF) { (ndf, c) => filterOutNulls(ndf, c) }
+      collectionColumns.foldLeft(geoConvertedDF) { (ndf, c) => filterOutNulls(ndf, c) }
     } else {
-      dateConvertedDF
+      geoConvertedDF
     }
 
     // Convert date/timestamp columns to ISO8601 strings for Azure Search

--- a/cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/search/AzureSearch.scala
+++ b/cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/search/AzureSearch.scala
@@ -258,18 +258,19 @@ object AzureSearchWriter extends IndexParser with IndexJsonGetter with SLogging 
    * (e.g. `{"type":"Point","coordinates":[lon, lat]}`), not as a JSON-encoded string.
    * Users frequently have their GeoJSON readily available as a string column, and
    * passing it as a `StringType` previously caused a `400 Bad Request`
-   * (see issue #2420) because the writer JSON-escaped the entire string.
+   * (see [[https://github.com/microsoft/SynapseML/issues/2420]]) because the writer
+   * JSON-escaped the entire string.
    *
    * For each field declared as `Edm.GeographyPoint` in the index, if the corresponding
    * DataFrame column is a `StringType`, parse it into the canonical
-   * `struct<type:string, coordinates:array<double>>` so that downstream `to_json`
-   * emits a proper GeoJSON object. Columns that are already structured are left as-is.
+   * `StructType(type: StringType, coordinates: ArrayType(DoubleType))` so that downstream
+   * `to_json` emits a proper GeoJSON object. Columns that are already structured are left as-is.
    *
    * @param df DataFrame with potential GeographyPoint columns
    * @param indexJson JSON string containing the index schema
    * @return DataFrame with string GeographyPoint columns converted to GeoJSON structs
    */
-  private def convertGeographyPointToStruct(df: DataFrame, indexJson: String): DataFrame = {
+  private[ml] def convertGeographyPointToStruct(df: DataFrame, indexJson: String): DataFrame = {
     val geoStructType = StructType(Seq(
       StructField("type", StringType),
       StructField("coordinates", ArrayType(DoubleType))
@@ -285,7 +286,7 @@ object AzureSearchWriter extends IndexParser with IndexJsonGetter with SLogging 
               when(col(fieldName).isNotNull, from_json(col(fieldName), geoStructType))
             )
           case _ =>
-            // Already a struct (or otherwise compatible) — let checkSchemaParity validate it.
+            // Already a struct (or otherwise compatible); checkSchemaParity will validate.
             currentDF
         }
       } else {

--- a/cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/search/AzureSearch.scala
+++ b/cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/search/AzureSearch.scala
@@ -261,10 +261,15 @@ object AzureSearchWriter extends IndexParser with IndexJsonGetter with SLogging 
    * (see [[https://github.com/microsoft/SynapseML/issues/2420]]) because the writer
    * JSON-escaped the entire string.
    *
-   * For each field declared as `Edm.GeographyPoint` in the index, if the corresponding
-   * DataFrame column is a `StringType`, parse it into the canonical
+   * For each '''top-level''' field declared as `Edm.GeographyPoint` in the index, if the
+   * corresponding DataFrame column is a `StringType`, parse it into the canonical
    * `StructType(type: StringType, coordinates: ArrayType(DoubleType))` so that downstream
-   * `to_json` emits a proper GeoJSON object. Columns that are already structured are left as-is.
+   * `to_json` emits a proper GeoJSON object. Columns that are already structured are
+   * left as-is. GeographyPoint fields nested inside complex types are not auto-converted
+   * (mirrors the existing top-level-only handling in `convertDateTimeToISO8601`).
+   *
+   * Parsing uses Spark's `FAILFAST` mode so malformed GeoJSON surfaces an explicit
+   * exception instead of being silently coerced to `null` and shipped to Azure Search.
    *
    * @param df DataFrame with potential GeographyPoint columns
    * @param indexJson JSON string containing the index schema
@@ -275,6 +280,7 @@ object AzureSearchWriter extends IndexParser with IndexJsonGetter with SLogging 
       StructField("type", StringType),
       StructField("coordinates", ArrayType(DoubleType))
     ))
+    val parseOptions = Map("mode" -> "FAILFAST")
     val geoFields = parseIndexJson(indexJson).fields
       .filter(_.`type` == "Edm.GeographyPoint")
       .map(_.name)
@@ -283,7 +289,7 @@ object AzureSearchWriter extends IndexParser with IndexJsonGetter with SLogging 
         currentDF.schema(fieldName).dataType match {
           case StringType =>
             currentDF.withColumn(fieldName,
-              when(col(fieldName).isNotNull, from_json(col(fieldName), geoStructType))
+              when(col(fieldName).isNotNull, from_json(col(fieldName), geoStructType, parseOptions))
             )
           case _ =>
             // Already a struct (or otherwise compatible); checkSchemaParity will validate.

--- a/cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/search/AzureSearch.scala
+++ b/cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/search/AzureSearch.scala
@@ -20,8 +20,8 @@ import org.apache.spark.ml.util._
 import org.apache.spark.ml.{ComplexParamsReadable, NamespaceInjections, PipelineModel}
 import org.apache.spark.ml.linalg.SQLDataTypes.VectorType
 import org.apache.spark.ml.functions.vector_to_array
-import org.apache.spark.sql.functions.{col, expr, from_json, struct, to_json, to_utc_timestamp,
-  date_format, when}
+import org.apache.spark.sql.functions.{col, concat, expr, forall, from_json, lit, raise_error, size,
+  struct, to_json, to_utc_timestamp, date_format, when}
 import org.apache.spark.sql.streaming.DataStreamWriter
 import org.apache.spark.sql.types._
 import org.apache.spark.sql.{DataFrame, Dataset, Row}
@@ -270,34 +270,52 @@ object AzureSearchWriter extends IndexParser with IndexJsonGetter with SLogging 
    *
    * Parsing uses Spark's `FAILFAST` mode so malformed GeoJSON surfaces an explicit
    * exception instead of being silently coerced to `null` and shipped to Azure Search.
+   * `FAILFAST` alone only rejects syntactically invalid JSON, so the parsed value is
+   * additionally validated to be a genuine GeoJSON Point (`type == "Point"` with exactly
+   * two non-null coordinates). Anything else raises an error naming the column and the
+   * offending value rather than indexing a silently-null location. NULL inputs are
+   * preserved as NULL.
    *
    * @param df DataFrame with potential GeographyPoint columns
    * @param indexJson JSON string containing the index schema
    * @return DataFrame with string GeographyPoint columns converted to GeoJSON structs
    */
   private[ml] def convertGeographyPointToStruct(df: DataFrame, indexJson: String): DataFrame = {
-    // Derived from edmTypeToSparkType so the parsed shape can never drift from the type
-    // checkSchemaParity expects for Edm.GeographyPoint
-    val geoStructType = edmTypeToSparkType(GeographyPointEdmType, None)
-    val parseOptions = Map("mode" -> "FAILFAST")
-    val geoFields = parseIndexJson(indexJson).fields
-      .filter(_.`type` == GeographyPointEdmType)
-      .map(_.name)
-    geoFields.foldLeft(df) { (currentDF, fieldName) =>
-      if (currentDF.columns.contains(fieldName)) {
-        currentDF.schema(fieldName).dataType match {
-          case StringType =>
-            currentDF.withColumn(fieldName,
-              when(col(fieldName).isNotNull, from_json(col(fieldName), geoStructType, parseOptions))
-            )
-          case _ =>
-            // Already a struct (or otherwise compatible); checkSchemaParity will validate.
-            currentDF
-        }
-      } else {
-        currentDF
-      }
-    }
+   // Derived from edmTypeToSparkType so the parsed shape can never drift from the type
+   // checkSchemaParity expects for Edm.GeographyPoint
+   val geoStructType = edmTypeToSparkType(GeographyPointEdmType, None)
+   val parseOptions = Map("mode" -> "FAILFAST")
+   val geoFields = parseIndexJson(indexJson).fields
+     .filter(_.`type` == GeographyPointEdmType)
+     .map(_.name)
+   geoFields.foldLeft(df) { (currentDF, fieldName) =>
+     if (currentDF.columns.contains(fieldName)) {
+       currentDF.schema(fieldName).dataType match {
+         case StringType =>
+           val parsed = from_json(col(fieldName), geoStructType, parseOptions)
+           val coordinates = parsed.getField("coordinates")
+           val isValidPoint = parsed.getField("type") === lit("Point") &&
+             coordinates.isNotNull &&
+             size(coordinates) === lit(GeographyPointCoordinateCount) &&
+             forall(coordinates, c => c.isNotNull)
+           val invalidValueError = raise_error(concat(
+             lit(s"AzureSearchWriter: column '$fieldName' is mapped to an " +
+               s"$GeographyPointEdmType field but the value is not a valid GeoJSON Point " +
+               """(expected {"type":"Point","coordinates":[longitude,latitude]}). """ +
+               "Offending value: "),
+             col(fieldName)))
+           currentDF.withColumn(fieldName,
+             when(col(fieldName).isNull || isValidPoint, parsed)
+               .otherwise(invalidValueError.cast(geoStructType))
+           )
+         case _ =>
+           // Already a struct (or otherwise compatible); checkSchemaParity will validate.
+           currentDF
+       }
+     } else {
+       currentDF
+     }
+   }
   }
 
   private def dfToIndexJson(schema: StructType,
@@ -504,6 +522,9 @@ object AzureSearchWriter extends IndexParser with IndexJsonGetter with SLogging 
   }
 
   private[ml] val GeographyPointEdmType = "Edm.GeographyPoint"
+
+  // GeoJSON Points are always [longitude, latitude]
+  private[ml] val GeographyPointCoordinateCount = 2
 
   private[ml] def edmTypeToSparkType(dt: String,  //scalastyle:ignore cyclomatic.complexity
                                      fields: Option[Seq[IndexField]]): DataType = dt match {

--- a/cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/search/AzureSearch.scala
+++ b/cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/search/AzureSearch.scala
@@ -20,7 +20,8 @@ import org.apache.spark.ml.util._
 import org.apache.spark.ml.{ComplexParamsReadable, NamespaceInjections, PipelineModel}
 import org.apache.spark.ml.linalg.SQLDataTypes.VectorType
 import org.apache.spark.ml.functions.vector_to_array
-import org.apache.spark.sql.functions.{col, expr, struct, to_json, to_utc_timestamp, date_format, when}
+import org.apache.spark.sql.functions.{col, expr, from_json, struct, to_json, to_utc_timestamp,
+  date_format, when}
 import org.apache.spark.sql.streaming.DataStreamWriter
 import org.apache.spark.sql.types._
 import org.apache.spark.sql.{DataFrame, Dataset, Row}
@@ -249,6 +250,50 @@ object AzureSearchWriter extends IndexParser with IndexJsonGetter with SLogging 
     }
   }
 
+  /**
+   * Converts string columns containing GeoJSON to the proper struct shape required for
+   * Azure Search `Edm.GeographyPoint` fields.
+   *
+   * Azure AI Search expects spatial values to be sent as a GeoJSON object
+   * (e.g. `{"type":"Point","coordinates":[lon, lat]}`), not as a JSON-encoded string.
+   * Users frequently have their GeoJSON readily available as a string column, and
+   * passing it as a `StringType` previously caused a `400 Bad Request`
+   * (see issue #2420) because the writer JSON-escaped the entire string.
+   *
+   * For each field declared as `Edm.GeographyPoint` in the index, if the corresponding
+   * DataFrame column is a `StringType`, parse it into the canonical
+   * `struct<type:string, coordinates:array<double>>` so that downstream `to_json`
+   * emits a proper GeoJSON object. Columns that are already structured are left as-is.
+   *
+   * @param df DataFrame with potential GeographyPoint columns
+   * @param indexJson JSON string containing the index schema
+   * @return DataFrame with string GeographyPoint columns converted to GeoJSON structs
+   */
+  private def convertGeographyPointToStruct(df: DataFrame, indexJson: String): DataFrame = {
+    val geoStructType = StructType(Seq(
+      StructField("type", StringType),
+      StructField("coordinates", ArrayType(DoubleType))
+    ))
+    val geoFields = parseIndexJson(indexJson).fields
+      .filter(_.`type` == "Edm.GeographyPoint")
+      .map(_.name)
+    geoFields.foldLeft(df) { (currentDF, fieldName) =>
+      if (currentDF.columns.contains(fieldName)) {
+        currentDF.schema(fieldName).dataType match {
+          case StringType =>
+            currentDF.withColumn(fieldName,
+              when(col(fieldName).isNotNull, from_json(col(fieldName), geoStructType))
+            )
+          case _ =>
+            // Already a struct (or otherwise compatible) — let checkSchemaParity validate it.
+            currentDF
+        }
+      } else {
+        currentDF
+      }
+    }
+  }
+
   private def dfToIndexJson(schema: StructType,
                             indexName: String,
                             keyCol: String,
@@ -367,17 +412,18 @@ object AzureSearchWriter extends IndexParser with IndexJsonGetter with SLogging 
       SearchIndex.createIfNoneExists(auth, serviceName, indexJson, apiVersion)
     }
     val dateConvertedDF = convertDateTimeToISO8601(preppedDF, indexJson)
+    val geoConvertedDF = convertGeographyPointToStruct(dateConvertedDF, indexJson)
 
     logInfo("checking schema parity")
-    checkSchemaParity(dateConvertedDF.schema, indexJson, actionCol)
+    checkSchemaParity(geoConvertedDF.schema, indexJson, actionCol)
 
     val df1 = if (filterNulls) {
       val collectionColumns = parseIndexJson(indexJson).fields
         .filter(_.`type`.startsWith("Collection"))
         .map(_.name)
-      collectionColumns.foldLeft(dateConvertedDF) { (ndf, c) => filterOutNulls(ndf, c) }
+      collectionColumns.foldLeft(geoConvertedDF) { (ndf, c) => filterOutNulls(ndf, c) }
     } else {
-      dateConvertedDF
+      geoConvertedDF
     }
 
     // Convert date/timestamp columns to ISO8601 strings for Azure AI Search

--- a/cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/search/AzureSearch.scala
+++ b/cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/search/AzureSearch.scala
@@ -276,13 +276,12 @@ object AzureSearchWriter extends IndexParser with IndexJsonGetter with SLogging 
    * @return DataFrame with string GeographyPoint columns converted to GeoJSON structs
    */
   private[ml] def convertGeographyPointToStruct(df: DataFrame, indexJson: String): DataFrame = {
-    val geoStructType = StructType(Seq(
-      StructField("type", StringType),
-      StructField("coordinates", ArrayType(DoubleType))
-    ))
+    // Derived from edmTypeToSparkType so the parsed shape can never drift from the type
+    // checkSchemaParity expects for Edm.GeographyPoint
+    val geoStructType = edmTypeToSparkType(GeographyPointEdmType, None)
     val parseOptions = Map("mode" -> "FAILFAST")
     val geoFields = parseIndexJson(indexJson).fields
-      .filter(_.`type` == "Edm.GeographyPoint")
+      .filter(_.`type` == GeographyPointEdmType)
       .map(_.name)
     geoFields.foldLeft(df) { (currentDF, fieldName) =>
       if (currentDF.columns.contains(fieldName)) {
@@ -504,6 +503,8 @@ object AzureSearchWriter extends IndexParser with IndexJsonGetter with SLogging 
     t.substring("Collection(".length).dropRight(1)
   }
 
+  private[ml] val GeographyPointEdmType = "Edm.GeographyPoint"
+
   private[ml] def edmTypeToSparkType(dt: String,  //scalastyle:ignore cyclomatic.complexity
                                      fields: Option[Seq[IndexField]]): DataType = dt match {
     case t if isEdmCollection(t) =>
@@ -515,7 +516,7 @@ object AzureSearchWriter extends IndexParser with IndexJsonGetter with SLogging 
     case "Edm.Double" => DoubleType
     case "Edm.Single" => FloatType
     case "Edm.DateTimeOffset" => StringType // We convert date/time to ISO8601 strings
-    case "Edm.GeographyPoint"   =>
+    case GeographyPointEdmType =>
       StructType(Seq(
         StructField("type", StringType),
         StructField("coordinates", ArrayType(DoubleType))

--- a/cognitive/src/test/scala/com/microsoft/azure/synapse/ml/services/search/split2/SearchWriterSuitePart2.scala
+++ b/cognitive/src/test/scala/com/microsoft/azure/synapse/ml/services/search/split2/SearchWriterSuitePart2.scala
@@ -204,4 +204,63 @@ class SearchWriterSuite extends SearchWriterSuiteUtilities {
 
   }
 
+  test("convertGeographyPointToStruct parses GeoJSON strings into structs") {
+    val df = spark.createDataFrame(Seq(
+      ("0", """{"type":"Point","coordinates":[-122.3493, 47.6205]}"""),
+      ("1", null)
+    )).toDF("id", "location")
+
+    val indexJson =
+      """
+        |{
+        |  "name": "unit-test-geo",
+        |  "fields": [
+        |    { "name": "id", "type": "Edm.String", "key": true },
+        |    { "name": "location", "type": "Edm.GeographyPoint" }
+        |  ]
+        |}
+        |""".stripMargin
+
+    val converted = AzureSearchWriter.convertGeographyPointToStruct(df, indexJson)
+    val expected = StructType(Seq(
+      StructField("type", StringType),
+      StructField("coordinates", ArrayType(DoubleType))
+    ))
+    assert(converted.schema("location").dataType == expected)
+
+    val rows = converted.orderBy("id").collect()
+    val parsed = rows.head.getStruct(rows.head.fieldIndex("location"))
+    assert(parsed.getString(0) == "Point")
+    assert(parsed.getSeq[Double](1) == Seq(-122.3493, 47.6205))
+    assert(rows(1).isNullAt(rows(1).fieldIndex("location")))
+  }
+
+  test("convertGeographyPointToStruct leaves struct columns untouched") {
+    val schema = StructType(Seq(
+      StructField("id", StringType),
+      StructField("location", StructType(Seq(
+        StructField("type", StringType, nullable = false),
+        StructField("coordinates", ArrayType(DoubleType, containsNull = false), nullable = false)
+      )))
+    ))
+    val df = spark.createDataFrame(
+      spark.sparkContext.parallelize(Seq(Row("0", Row("Point", Seq(-122.3493, 47.6205))))),
+      schema
+    )
+
+    val indexJson =
+      """
+        |{
+        |  "name": "unit-test-geo",
+        |  "fields": [
+        |    { "name": "id", "type": "Edm.String", "key": true },
+        |    { "name": "location", "type": "Edm.GeographyPoint" }
+        |  ]
+        |}
+        |""".stripMargin
+
+    val converted = AzureSearchWriter.convertGeographyPointToStruct(df, indexJson)
+    assert(converted.schema("location").dataType == schema("location").dataType)
+  }
+
 }

--- a/cognitive/src/test/scala/com/microsoft/azure/synapse/ml/services/search/split2/SearchWriterSuitePart2.scala
+++ b/cognitive/src/test/scala/com/microsoft/azure/synapse/ml/services/search/split2/SearchWriterSuitePart2.scala
@@ -200,6 +200,10 @@ class SearchWriterSuite extends SearchWriterSuiteUtilities {
       )
     )
 
+    // With fatalErrors=true (default) any 400 from Azure Search becomes a thrown
+    // RuntimeException, so reaching this `assertSize` proves the documents were
+    // accepted as valid spatial objects -- a count of 2 is only achievable if the
+    // GeoJSON strings were correctly parsed and serialized as GeoJSON objects.
     retryWithBackoff(assertSize(in, 2))
 
   }
@@ -261,6 +265,29 @@ class SearchWriterSuite extends SearchWriterSuiteUtilities {
 
     val converted = AzureSearchWriter.convertGeographyPointToStruct(df, indexJson)
     assert(converted.schema("location").dataType == schema("location").dataType)
+  }
+
+  test("convertGeographyPointToStruct fails fast on malformed GeoJSON instead of silently nulling") {
+    val df = spark.createDataFrame(Seq(
+      ("0", "{not valid json")
+    )).toDF("id", "location")
+
+    val indexJson =
+      """
+        |{
+        |  "name": "unit-test-geo",
+        |  "fields": [
+        |    { "name": "id", "type": "Edm.String", "key": true },
+        |    { "name": "location", "type": "Edm.GeographyPoint" }
+        |  ]
+        |}
+        |""".stripMargin
+
+    val converted = AzureSearchWriter.convertGeographyPointToStruct(df, indexJson)
+    // FAILFAST surfaces parse errors when the row is materialized, not at plan time.
+    intercept[org.apache.spark.SparkException] {
+      converted.collect()
+    }
   }
 
 }

--- a/cognitive/src/test/scala/com/microsoft/azure/synapse/ml/services/search/split2/SearchWriterSuitePart2.scala
+++ b/cognitive/src/test/scala/com/microsoft/azure/synapse/ml/services/search/split2/SearchWriterSuitePart2.scala
@@ -171,4 +171,37 @@ class SearchWriterSuite extends SearchWriterSuiteUtilities {
 
   }
 
+  test("Handle GeoJSON GeographyPoint fields supplied as strings") {
+
+    val in = generateIndexName()
+    val df = spark.createDataFrame(Seq(
+      ("upload", "0", """{"type":"Point","coordinates":[-122.3493, 47.6205]}"""),
+      ("upload", "1", """{"type":"Point","coordinates":[-122.3351, 47.6080]}""")
+    )).toDF("searchAction", "id", "location")
+
+    val indexJson =
+      s"""
+         |{
+         |  "name": "$in",
+         |  "fields": [
+         |    { "name": "id", "type": "Edm.String", "key": true, "searchable": true, "retrievable": true },
+         |    { "name": "location", "type": "Edm.GeographyPoint", "searchable": false,
+         |     "filterable": true, "retrievable": true, "sortable": true }
+         |  ]
+         |}
+         |""".stripMargin
+
+    AzureSearchWriter.write(df,
+      Map(
+        "subscriptionKey" -> azureSearchKey,
+        "actionCol" -> "searchAction",
+        "serviceName" -> testServiceName,
+        "indexJson" -> indexJson
+      )
+    )
+
+    retryWithBackoff(assertSize(in, 2))
+
+  }
+
 }

--- a/cognitive/src/test/scala/com/microsoft/azure/synapse/ml/services/search/split2/SearchWriterSuitePart2.scala
+++ b/cognitive/src/test/scala/com/microsoft/azure/synapse/ml/services/search/split2/SearchWriterSuitePart2.scala
@@ -285,9 +285,83 @@ class SearchWriterSuite extends SearchWriterSuiteUtilities {
 
     val converted = AzureSearchWriter.convertGeographyPointToStruct(df, indexJson)
     // FAILFAST surfaces parse errors when the row is materialized, not at plan time.
-    intercept[org.apache.spark.SparkException] {
+    // The concrete wrapper type varies (SparkException, ExecutionException, or a bare
+    // RuntimeException when Spark folds the LocalRelation on the driver), so assert on
+    // the flattened cause chain instead.
+    val caught = intercept[Exception] {
       converted.collect()
     }
+    assert(causeChain(caught).contains("Malformed records are detected"),
+      s"expected a FAILFAST parse failure but got: ${causeChain(caught)}")
   }
+
+  test("convertGeographyPointToStruct rejects valid JSON that is not a GeoJSON Point") {
+    val indexJson =
+      """
+        |{
+        |  "name": "unit-test-geo",
+        |  "fields": [
+        |    { "name": "id", "type": "Edm.String", "key": true },
+        |    { "name": "location", "type": "Edm.GeographyPoint" }
+        |  ]
+        |}
+        |""".stripMargin
+
+    // Every one of these is syntactically valid JSON (or blank), so Spark's FAILFAST parser
+    // accepts it and yields a partially-null struct. Without explicit shape validation these
+    // would be silently indexed as a null location instead of failing the write.
+    val wrongShapes = Seq(
+      """{"foo":"bar"}""",
+      """{"type":"Point"}""",
+      """{"coordinates":[-122.3493, 47.6205]}""",
+      """{"type":"Polygon","coordinates":[-122.3493, 47.6205]}""",
+      """{"type":"Point","coordinates":[-122.3493]}""",
+      """{"type":"Point","coordinates":[-122.3493, 47.6205, 12.0]}""",
+      """{"type":"Point","coordinates":[-122.3493, null]}""",
+      "",
+      "   "
+    )
+
+    wrongShapes.foreach { badValue =>
+      val df = spark.createDataFrame(Seq(("0", badValue))).toDF("id", "location")
+      val converted = AzureSearchWriter.convertGeographyPointToStruct(df, indexJson)
+      val caught = intercept[Exception] {
+        converted.collect()
+      }
+      val message = causeChain(caught)
+      assert(message.contains("not a valid GeoJSON Point"),
+        s"expected a GeoJSON validation failure for '$badValue' but got: $message")
+      assert(message.contains("location"),
+        s"expected the error to name the offending column for '$badValue'")
+    }
+  }
+
+  test("convertGeographyPointToStruct preserves nulls without raising") {
+    val df = spark.createDataFrame(Seq(
+      ("0", """{"type":"Point","coordinates":[-122.3493, 47.6205]}"""),
+      ("1", null),
+      ("2", null)
+    )).toDF("id", "location")
+
+    val indexJson =
+      """
+        |{
+        |  "name": "unit-test-geo",
+        |  "fields": [
+        |    { "name": "id", "type": "Edm.String", "key": true },
+        |    { "name": "location", "type": "Edm.GeographyPoint" }
+        |  ]
+        |}
+        |""".stripMargin
+
+    val rows = AzureSearchWriter.convertGeographyPointToStruct(df, indexJson).orderBy("id").collect()
+    assert(rows.length == 3)
+    assert(!rows.head.isNullAt(rows.head.fieldIndex("location")))
+    assert(rows(1).isNullAt(rows(1).fieldIndex("location")))
+    assert(rows(2).isNullAt(rows(2).fieldIndex("location")))
+  }
+
+  private def causeChain(t: Throwable): String =
+    Iterator.iterate(t)(_.getCause).takeWhile(_ != null).map(_.toString).mkString(" | ")
 
 }

